File: skills/webhook-consumer-architecture/examples/python/django/validation.py
from datetime import datetime

from pydantic import BaseModel


class ProviderSubscriptionPayload(BaseModel):
    id: str
    customer_id: str
    plan_id: str
    status: str
    current_term_end: datetime | None = None


class NormalizedWebhookEvent(BaseModel):
    provider: str
    event_id: str
    event_type: str
    resource_id: str
    payload: dict


def normalize_subscription_event(data: dict, event_id: str, event_type: str) -> NormalizedWebhookEvent:
    payload = ProviderSubscriptionPayload.model_validate(data)

    return NormalizedWebhookEvent(
        provider="chargebee",
        event_id=event_id,
        event_type=event_type,
        resource_id=payload.id,
        payload=payload.model_dump(),
    )

File: skills/webhook-consumer-architecture/examples/go/net-http/validation.go
package examples

type ProviderSubscriptionPayload struct {
	ID         string `json:"id"`
	CustomerID string `json:"customer_id"`
	PlanID     string `json:"plan_id"`
	Status     string `json:"status"`
}

type NormalizedWebhookEvent struct {
	Provider   string
	EventID    string
	EventType  string
	ResourceID string
	Payload    map[string]any
}

func NormalizeSubscriptionEvent(payload ProviderSubscriptionPayload, eventID, eventType string) NormalizedWebhookEvent {
	return NormalizedWebhookEvent{
		Provider:   "chargebee",
		EventID:    eventID,
		EventType:  eventType,
		ResourceID: payload.ID,
		Payload: map[string]any{
			"customer_id": payload.CustomerID,
			"plan_id":     payload.PlanID,
			"status":      payload.Status,
		},
	}
}
